### PR TITLE
Chore: Remove lodash

### PR DESCRIPTION
**Pull request by stephenwade**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain:

This addresses part of #14098, reducing dependency file count by removing `lodash` from the project.

(Or, at least it should. Unfortunately, the `table` package (used by the table formatter) also depends on `lodash`. If the maintainers would consider deprecating the `table` formatter or splitting it into a separate package, that would finish removing `lodash` from the project. See https://github.com/eslint/eslint/issues/14098#issuecomment-782158309.)

#### What changes did you make? (Give an overview)

I removed all uses of lodash in the project, replacing them with custom code or other NPM packages.

#### Is there anything you'd like reviewers to focus on?

Nothing in particular. If you'd like me to split this into multiple PRs, I am happy to do so.

### CODE REVIEW

**Review by mdjermanovic — COMMENTED** [review]
> Unfortunately, the table package (used by the table formatter) also depends on lodash.

Then it looks like we shouldn't merge this change now as it would add more dependencies while still not removing `lodash` from the dependency tree?

> If the maintainers would consider deprecating the table formatter or splitting it into a separate package, that would finish removing lodash from the project. See #14098 (comment).

An alternative could be to ask `table` to remove `lodash`. It seems they're willing to do that, but had to [revert](https://github.com/gajus/table/pull/121) the first attempt.

File: lib/linter/apply-disable-directives.js [hunk 1]
```diff
@@ -124,7 +122,7 @@ module.exports = ({ directives, problems, reportUnusedDisableDirectives = "off"
         .map(directive => Object.assign({}, directive, { unprocessedDirective: directive }))
         .sort(compareLocations);
 
-    const lineDirectives = lodash.flatMap(directives, directive => {
+    const lineDirectives = directives.flatMap(directive => {
```
Comment (mdjermanovic) [inline lib/linter/apply-disable-directives.js:125]: Per [node.green](https://node.green/), `Array.prototype.flatMap` is available on Node >= 12.4, while ESLint 7 supports these versions:

https://github.com/eslint/eslint/blob/ebfb63a682004a008f2707dbad616e5ae1630b2c/package.json#L144-L146

so this change would have to wait for the next ESLint major release.

**Review by TimvdLippe — COMMENTED** [review]
Thank you so much going through these! In terms of process, it might be better to split up this PR into two parts:

1. A PR that removes usages of `lodash` and replaces them with idiomatic ES6 (given that the Node versions support it)
2. A PR that replaces usages of `lodash` with smaller dependencies

For example, in PR 1 you would be able to put the following refactorings (non-inclusive list, I quickly skimmed this PR):

1. `forEach`
2. `last`
3. `find`
4. `map`
5. `flat`
6. `noop`
7. `endsWith`
8. `findLast`
9. `includes`
10. `isEmpty`

For `memoize`, you might be able to use the same solution as I proposed in https://github.com/eslint/eslintrc/pull/29

All other changes would go into PR 2 and would be blocked on the `table` PR as well. The reason for splitting up is that it makes introducing new usages of `lodash` less likely, when idiomatic ES6 is an option. Hopefully that can improve the future maintainability of ESLint, by preventing backsliding "since we have `lodash` available anyways".

P.S. Not an ESLint maintainer, so they might have a differing opinion on this process proposal.

**Comment by mdjermanovic** [comment]
Replacing `lodash` functions with available native equivalents is probably something we should do regardless of removing `lodash`. That subset of changes looks like the best candidate for PR 1.

I think that replacing `lodash` functions with our custom code doesn't provide value if we're not sure that we can completely remove `lodash`, so that would probably go into PR 2, along with the new dependencies.

**Comment by stephenwade** [comment]
> Replacing lodash functions with available native equivalents is probably something we should do regardless of removing lodash. That subset of changes looks like the best candidate for PR 1.

Okay, I'll split the PR up into two, and I'll make sure to test on Node 10 as well as my default Node 15.

**Pull request closed without merging**